Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabPublicHandler.cpp

```diff
@@ -1311,10 +1283,116 @@ namespace AzToolsFramework
+                    PrefabDomPath parentEntityAliasInOwningTemplatePath(parentEntityAlias.c_str());
+                    PrefabDomValue* parentEntityDomPtr = parentEntityAliasInOwningTemplatePath.Get(commonOwningTemplateDom);
+                    if (parentEntityDomPtr)
+                    {
+                        // Deep copy is needed to capture the current DOM state.
+                        parentEntityDomBeforeRemovalMap[parentEntityId] = PrefabDom();
+                        PrefabDom& domInMap = parentEntityDomBeforeRemovalMap[parentEntityId];
+                        domInMap.CopyFrom(*parentEntityDomPtr, domInMap.GetAllocator());
+                    }
+                }
+
+                // Detaches the instance, removes the link and instance DOM in template.
+                AZStd::unique_ptr<Instance> outInstance =
+                    commonOwningInstance->get().DetachNestedInstance(nestedInstance->GetInstanceAlias());
+                RemoveLink(outInstance, commonOwningTemplateId, undoBatch.GetUndoBatch());
+                outInstance.reset();
+            }
+
+            // Step 2 - Entity Removals.
+            AZStd::vector<PrefabDom> removedEntityDomList;
+            AZStd::vector<AZStd::string> removedEntityAliasPathList;
+            AZStd::unordered_set<AZ::EntityId> entitiesThatGotRemoved;
+
+            for (const AZ::Entity* entity : entityList)
+            {
+                const AZ::EntityId entityId = entity->GetId();
+                const AZStd::string entityAliasPath = m_instanceToTemplateInterface->GenerateEntityAliasPath(entityId);
+
+                // If parent is valid and not seen before (because we only need to capture the initial state),
+                // then get the before state of the parent entity DOM.
+                AZ::EntityId parentEntityId;
+                AZ::TransformBus::EventResult(parentEntityId, entityId, &AZ::TransformBus::Events::GetParentId);
+                if (parentEntityId.IsValid() &&
+                    parentEntityDomBeforeRemovalMap.find(parentEntityId) == parentEntityDomBeforeRemovalMap.end())
+                {
+                    AZStd::string parentEntityAlias = m_instanceToTemplateInterface->GenerateEntityAliasPath(parentEntityId);
+                    PrefabDomPath parentEntityAliasInOwningTemplatePath(parentEntityAlias.c_str());
+                    PrefabDomValue* parentEntityDomPtr = parentEntityAliasInOwningTemplatePath.Get(commonOwningTemplateDom);
+                    if (parentEntityDomPtr)
+                    {
+                        // Deep copy is needed to capture the current DOM state.
+                        parentEntityDomBeforeRemovalMap[parentEntityId] = PrefabDom();
+                        PrefabDom& domInMap = parentEntityDomBeforeRemovalMap[parentEntityId];
+                        domInMap.CopyFrom(*parentEntityDomPtr, domInMap.GetAllocator());
+                    }
```

Comment:
Accessing the value immediately after insertion can be avoided if you do
```
PrefabDom parentEntityDomBeforeRemoval;
ParentEntityDomBeforeRemoval.CopyFrom();
//insert into map
```

Reply:
This is actually what I was doing.

```cpp
PrefabDom parentEntityDomBeforeRemoval;
ParentEntityDomBeforeRemoval.CopyFrom();
parentEntityDomBeforeRemovalMap[parentEntityId] = parentEntityDomBeforeRemoval;
```

But can the copying in line 3 be avoided? In other words, we create the PrefabDom in place instead by doing `parentEntityDomBeforeRemovalMap.emplace(parentEntityId)`.

Reply:
You  can use AZStd::move to move the newly created dom into the map. That'll skip the copy.